shared/src/http.rs

```diff
@@ -131,6 +135,109 @@ pub fn filter_hop_by_hop(headers: &mut HeaderMap, version: Version) -> &mut Head
     headers
 }
 
+/// Send a request to a single upstream with configurable timeout
+pub async fn send_to_upstream<C, B, E>(
+    client: &Client<C, B>,
+    upstream_url: &url::Url,
+    request: Request<B>,
+    timeout_secs: u64,
+) -> Result<Response<Bytes>, E>
+where
+    C: hyper_util::client::legacy::connect::Connect + Clone + Send + Sync + 'static,
+    B: hyper::body::Body + Send + Unpin + 'static,
+    B::Data: Send,
+    B::Error: std::error::Error + Send + Sync + 'static,
+    E: From<UpstreamError>,
+{
+    // Use host as identifier for error messages
+    let upstream_identifier = upstream_url.host_str().unwrap_or(upstream_url.as_str());
+
+    // Build the full upstream URI by combining base URL with request path
+    let path_and_query = request
+        .uri()
+        .path_and_query()
+        .map(|pq| pq.as_str())
+        .unwrap_or("/");
+
+    let mut url = upstream_url.clone();
+    if let Some((path, query)) = path_and_query.split_once('?') {
+        url.set_path(path);
+        url.set_query(Some(query));
+    } else {
+        url.set_path(path_and_query);
+    }
+    let upstream_uri = url.to_string();
+
+    // Build request to send to upstream with modified URI and filtered headers
+    let (mut parts, body) = request.into_parts();
+    filter_hop_by_hop(&mut parts.headers, parts.version);
+
+    let mut req_builder = Request::builder()
+        .method(parts.method)
+        .uri(upstream_uri)
+        .version(parts.version);
+
+    for (name, value) in parts.headers.iter() {
+        req_builder = req_builder.header(name, value);
+    }
+
+    let upstream_request = req_builder.body(body).map_err(|e| {
+        E::from(UpstreamError::InternalError(format!(
+            "Failed to build request: {e}"
+        )))
+    })?;
+
+    // Send request with timeout
+    let response = timeout(
+        Duration::from_secs(timeout_secs),
+        client.request(upstream_request),
+    )
+    .await
+    .map_err(|_| E::from(UpstreamError::Timeout(upstream_identifier.to_string())))?
+    .map_err(|e| {
+        E::from(UpstreamError::RequestFailed(
+            upstream_identifier.to_string(),
+            e.to_string(),
+        ))
+    })?;
+
+    // Collect response body bytes
+    let (parts, body) = response.into_parts();
+    let body_bytes = body
+        .collect()
+        .await
+        .map(|collected| collected.to_bytes())
+        .map_err(|e| E::from(UpstreamError::ResponseBodyError(e.to_string())))?;
+
+    Ok(Response::from_parts(parts, body_bytes))
+}
+
+/// Errors that can occur during upstream operations
+#[derive(Debug)]
+pub enum UpstreamError {
+    RequestBodyError(String),
+    ResponseBodyError(String),
+    Timeout(String),
+    RequestFailed(String, String),
+    InternalError(String),
+}
+
+impl std::fmt::Display for UpstreamError {
```

Comment:
can we use thiserror to match the rest of the errors in synapse + don't have to implement our own Display